src/java/com/google/cloud/bigquery/dwhassessment/extractiontool/executor/SaveCheckerImpl.java

```diff
@@ -0,0 +1,165 @@
+  // respectively. See go/chunked-dwh-assessment-extraction-dd for further details.
+  private static final Pattern INPUT_CHUNK_PATTERN =
+      Pattern.compile("([\\w_]+)-(\\d{8}T\\d{6}S\\d{6})-(\\d{8}T\\d{6}S\\d{6})_(\\d+)\\.avro");
+
+  // TODO(cyulysses-corp): collect different date-time parser/formatters into the same place for
+  // reliable references.
+  private static final DateTimeFormatter chunkTimestampFormatter =
+      new DateTimeFormatterBuilder()
+          .parseCaseSensitive()
+          .append(DateTimeFormatter.BASIC_ISO_DATE)
+          .appendLiteral("T")
+          .append(DateTimeFormatter.ofPattern("HHmmss"))
+          .appendLiteral("S")
+          .append(DateTimeFormatter.ofPattern("SSSSSS"))
+          .toFormatter(Locale.ENGLISH);
+
+  public SaveCheckerImpl(ImmutableMap<String, ImmutableList<String>> sortingColumnsMap) {
+    this.sortingColumnsMap = sortingColumnsMap;
+  }
+
+  @Override
+  public ImmutableMap<String, ChunkCheckpoint> getScriptCheckPoints(Path path) {
+    Map<String, List<Matcher>> fileMap;
+    try {
+      fileMap =
+          Files.walk(path)
+              .filter(Files::isRegularFile)
+              .map(oneFile -> INPUT_CHUNK_PATTERN.matcher(oneFile.getFileName().toString()))
+              .filter(Matcher::matches)
+              .collect(
+                  groupingBy(
+                      (Matcher matcher) -> matcher.group(1),
+                      mapping(
+                          Function.identity(),
+                          collectingAndThen(
+                              toList(),
+                              list ->
+                                  list.stream()
+                                      .sorted(
+                                          Comparator.comparingInt(
+                                              matcher -> Integer.parseInt(matcher.group(4))))
+                                      .collect(toList())))));
+    } catch (IOException e) {
+      throw new IllegalStateException(
+          String.format("Error reading path '%s'.", path.toString()), e);
+    }
+    ImmutableMap.Builder<String, ChunkCheckpoint> checkPointsMapBuilder = ImmutableMap.builder();
+    for (String scriptName : sortingColumnsMap.keySet()) {
+      if (fileMap.containsKey(scriptName) && !fileMap.get(scriptName).isEmpty()) {
+        List<Matcher> matchers = fileMap.get(scriptName);
+        // Note that the validator only validates information contained in the file names.
+        // The timestamp comparisons are only done lexicographically without checking whether each
+        // timestamp can be parsed. This is because the probability of having incorrectly generated
+        // timestamps in filenames while other aspects are normal is very low.
+        validateSortedChunkSequence(matchers);
+        Matcher lastMatcher = Iterables.getLast(matchers);
+        Instant lastInstant;
+        try {
+          lastInstant =
+              ZonedDateTime.of(
```

Comment:
Can we extract `getUtcInstantFromDatetimeAndZone` to utils and reuse?

Reply:
Made it a private method. If it is to be reused, as I mentioned in TODO, we'd better have a well-designed util class dedicated to unifying the different time representations.